imsim/stamp.py

```diff
@@ -88,32 +87,37 @@ def setup(self, config, base, xsize, ysize, ignore, logger):
         # First do a parsing check to make sure that all the options passed to
         # the config are valid, and no required options are missing.
 
-        req = {'diffraction_psf': dict, 'det_name': str}
-        opt = {'camera': str}
-        # For the optional ones we don't need here, can put in ignore, and they will still
-        # be allowed, but not required.
-        ignore = ['fft_sb_thresh', 'band', 'airmass', 'rawSeeing', 'max_flux_simple'] + ignore
+        try:
+            self.vignetting = GetInputObj('vignetting', config, base, 'LSST_SiliconBuilder')
+        except galsim.config.GalSimConfigError:
+            self.vignetting = None
+
+        req = {}
+        opt = {'camera': str, 'diffraction_fft': dict,
+               'airmass': float, 'rawSeeing': float, 'band': str}
+        if self.vignetting:
+            req['det_name'] = str
+        else:
+            opt['det_name'] = str
+        # For the optional ones we parse manually, we can put in ignore, and they will
+        # still be allowed, but not required.
+        ignore = ['fft_sb_thresh', 'max_flux_simple'] + ignore
         params = galsim.config.GetAllParams(config, base, req=req, opt=opt, ignore=ignore)[0]
 
         gal = galsim.config.BuildGSObject(base, 'gal', logger=logger)[0]
         if gal is None:
             raise galsim.config.SkipThisObject('gal is None (invalid parameters)')
         self.gal = gal
 
-        self.diffraction_psf = DiffractionPSF.from_config(config, base)
-
-        try:
-            self.vignetting = GetInputObj('vignetting', config, base,
-                                          'LSST_SiliconBuilder')
-        except galsim.config.GalSimConfigError:
-            self.vignetting = None
+        self.diffraction_fft = DiffractionFFT.from_config(config, base)
```

Comment:
I went through this, and I think it is correct.  But I found it quite confusing when first reading to see the checks above and then a 2nd check using the dataclass here. 

That's because this diffraction_fft object is different, with it's own options and isn't actually used until the draw() function later. 

I think some comments saying something about also check the difraction_fft options for use later in fcn() might be useful.

Reply:
It's also using a sub-dict.  The GetAllParams that it calls isn't the same field as we're doing here.  It's config['diffraction_fft'].  So we couldn't include those parameters in the GetAllParams call we do at this point.

Reply:
Actually, I think a better API would have `from_config` take this sub-dict as it's parameter rather than the parent dict.  That would probably make it clearer here why this does a second round of GetAllParams.  I'll switch tot hat.